### ISSUE TITLE
Confusing behavior of extension types and iterable extensions

### ISSUE THREAD

**Issue by incendial** [dart-lang/sdk#55007]
Consider the following example:

```dart
// Copied from package:collection
extension _IterableNullableExtension<T extends Object> on Iterable<T?> {
  Iterable<T> whereNotNull() sync* {
    for (var element in this) {
      if (element != null) yield element;
    }
  }
}

extension type ET(String? value) {}

void fn() {
  final et = ET(null);
  final nullableTypes = [et, et] as Iterable<ET>;
  nullableTypes.whereNotNull(); // <-- analyzer error here
}
```

Produces the following error (tested on dartpad main channel):
<img width="526" alt="Screenshot 2024-02-24 at 16 46 08" src="https://github.com/dart-lang/sdk/assets/21116742/93217a0b-4475-460b-bcbc-9f60c95e35dd">

I'm not sure how this supposed to work, so here are several approaches I tried so far:
1) updating the extension `T` to extend `ET` instead of `Object`. This does work, but in the real-world code updating external extensions is not possible.
2) adding `implements Object` to ET. This does remove the analyzer error, but it requires the representation type to be non-nullable and shows (implementing Object? is not possible):
<img width="637" alt="Screenshot 2024-02-24 at 16 50 24" src="https://github.com/dart-lang/sdk/assets/21116742/6f647eb5-ecb8-4b3c-913e-5f9169440725">

But both of them require code change.

Can you advise the correct approach to resolve this error? At this point I'm not sure whether this behavior is by design or a bug.

Sidenote: the idea that both ET and the representation type can be marked as nullable separately (e.g. a function can accept ET with a nullable representation type or ET? with regular representation type) is also somewhat confusing.

**Comment by lrhn** [comment]
The behavior is correct. The type `ET` is not a subtype of `Object`, so there is no solution for `<T extends Object> on Iterable<T?>` on `List<ET>`.

The solutions are what you already mention:
- Change `ET` to have `String` as representation type, and implement `Object`, then use `ET?` as the element type of them list.
- Change the extension to `<T extends Object?> on Iterable<T?>`. That may give worse typing though.
- Or cast before calling, `([et, et] as List<String?>).nonNulls as List<ET>`.

The ET type is tricky because it's not non-nullable, and it's not nullable.
That means it may contain null, but it has no subtype where that `null` is excluded.

The more direct solution would be to use `.where((e) => e != null)`, which does not try to improve the type, because that's what fails.

**Comment by incendial** [comment]
Thanks!

Do I get it right that at runtime both `ET1(String? value)` and `ET2(String value)` (but used as ET2?) represent the same value `String?`?

**Comment by incendial** [comment]
Thanks! (it's funny how github shows edits in real-time 😁)

One more question, if I may: what's the reason for extension types not to implement Object implicitly? To allow `ET(String? value)`? Or there is something deeper to that decision?

**Comment by eernstg** [comment]
> what's the reason for extension types not to implement Object implicitly

When an extension type `E` has a non-extension type `T` as a superinterface, it is considered to be transparent with respect to that type. That is, it is being announced to clients that the representation type of `E` is some subtype of `T`, and in particular, expressions of type `E` can be used where a value of type `T` is required. (So we can "leave" the extension type silently).

```dart
extension type FancyString(String it) implements String {...}

void main() {
  FancyString fs = FancyString('Hello world!');
  String s = fs; // OK!
}
```

There is a [lint proposal](https://github.com/dart-lang/linter/issues/4760) which will flag run-time casts into an extension type _unless_ it is transparent with respect to the source type of the cast. Again, the transparent extension type is by convention expected to be an appropriate typing for _every_ object whose type is the representation type, and hence it's OK to obtain an expression of that type by a run-time cast (`Object o = ...; if (o is FancyString) ...`).

The clause `implements Object` is just the most extreme case, indicating that it's OK to cast from `Object` or a top type into this extension type.

At the same time, `implements Object` does imply that the representation type is non-nullable, which means that a test like `e == null` on an expression `e` with static type `E?` is known to admit every value of type `E`. If `E` has representation type `int?`, say, then an `E?` could have the value null because this is a possible value of the representation type, and it could have the value null because that's always a possibility for a type of the form `T?`. So we're avoiding the confusion with the "two reasons for being null" situation, which may be helpful when working with types like `E?`.

Conversely, if an extension type `E` does _not_ implement `Object` (hence, it doesn't implement any other non-extension type, either), then it is (again: by convention) implied that it is inappropriate to obtain a value of type `E` by a run-time type cast.

There's been a lot of controversy in the language team and in issues about this, but I certainly think it's worthwhile to help developers maintain this distinction.

**Comment by incendial** [comment]
Thanks for the detailed answer!

I'm currently working on supporting ET in DCM rules and got to the conclusion that we need a lint rule for casting to an extension type when it can be replaced with an invocation. And after seeing your last example (that covers that exact case), looks like this rule would be indeed useful.

I've also supported this:
```dart
extension type ET4(int? value) {}

void fn(String str) {
  '1' as ET4; // LINT, mismatching representation type
  if (str is ET4) { } // LINT, mismatching representation type
}
```
which should help avoiding some of the potential mistakes.

Closing this one.

**Comment by eernstg** [comment]
> Thanks for the detailed answer!

Thanks! One more thing, by the way:

```dart
extension type ET4(int? value) {}

void fn(String str) {
  '1' as ET4; // Identical to `'1' as int?` at run time.
  if (str is ET4) { } // Identical to `str is int?` at run time.
}
```

One might say that `'1' as ET4` and `'1' as int?` are similarly dangerous: They are guaranteed to throw at run time. So they should perhaps be linted equally (either both gets a lint, or none of them, because they are just convoluted ways to write `throw TypeError(...)`).

On the other hand, `1 as ET4` is dangerous for a different reason, even though it will succeed at run time (or perhaps _because_ it will succeed): It creates an expression of type `ET4` from an expression of type `int`, and the type `ET4` does not give us any reason to believe that this is an appropriate transition. That's exactly the kind of situation which will be linted by [cast_into_extionsion_type](https://github.com/dart-lang/linter/issues/4760).

The conceptual point is that a cast to a type which is reified at run time amounts to a _confirmation_ that the given object has that type, which is something that was chosen when that object was created. For example, every `String` was born to be a `String`, and there's nothing wrong in discovering that the value of a given expression of type `Object` is actually a `String`.

In contrast, a cast to a type which is not reified (in particular, an extension type) is a claim with no justification that the given object is an appropriate representation object for the given extension type. For example:

```dart
extension type Prime._(int it) {
  Prime(this.it) : assert(isPrime(it));
}

void main() {
  var p = 42 as Prime; // Accepted, skips the validation.
}
```

Extension types give us the opportunity to have certain compile-time checks, introducing distinctions between groups of objects with exactly the same representation at run time (in order words, the compile-time distinction has a zero cost at run time). However, dynamic type tests will eliminate this distinction. That's the reason why we should be careful about performing such dynamic type tests involving extension types.

For some extension types it's not a problem: They have been created under the assumption that _any_ object whose run-time type is a subtype of the representation type of the extension type is appropriate, all of them are good. In that case we can cast up and down, no problem. They are the extension types that I call 'transparent'; by convention, they are characterized by having an `implements` clause that includes a non-extension type, which is often the representation type (and must in any case be a supertype of the representation type).

**Comment by incendial** [comment]
> One might say that '1' as ET4 and '1' as int? are similarly dangerous:
> either both gets a lint

Completely agree.

>  I don't think there's any particular reason to lint '1' as ET4 because it's going to throw (unless we also want to lint '1' as int?)

Just to clarify, I'm not talking about the analyzer/linter rules, but about a separate tool. And we already provide a rule for the second case https://dcm.dev/docs/rules/common/avoid-unrelated-type-casts/, and this rule now also supports ETs.
<img width="654" alt="Screenshot 2024-02-28 at 19 59 00" src="https://github.com/dart-lang/sdk/assets/21116742/ed41f69f-e333-469e-8db4-fcf4a28b8d35">

> but I do think we might want to lint '1' as ET4 because it's a cast into an extension type that isn't transparent.

Did you mean `1 as ET4` (with int and not string)?

**Comment by incendial** [comment]
> because it's a cast into an extension type that isn't transparent.

And looks like even if ET is transparent, linting a cast can be useful (considering the number of ETs in the scope can be more than 1) as now there is no lint for:

```dart
extension type ET(String str) implements String {}

void fn() {
  '' as ET;
}
```

**Comment by eernstg** [comment]
>> but I do think we might want to lint `'1' as ET4` because it's a cast into an extension type that isn't transparent.
>
> Did you mean `1 as ET4` (with int and not string)? If not, I'm a bit confused by `I don't think there's any particular reason` and then `but I do think we might want to lint` 🤔

Right, there were some conflicting forces about the same syntax.

I did mean "we might want to lint `'1' as ET4` because it's a cast into an extension type that isn't transparent". This kind of cast would be linted by `cast_into_extension_type`, no matter whether it's going to throw. It's a separate question whether we'd also have a (hypothetical)`cast_always_throws` lint on the same expression. Of course, we'd _also_ lint `1 as ET4`.

As an aside, `cast_always_throws` is tricky, because most types can include values that are also subtypes of some other given type:

```dart
class A {}
class B {}
class C implements A, B {}

void main() {
  B b = C();
  if (b is A) print("Got an A!"); // Cast to unrelated type that succeeds.
}
```

Hence, it's a very special case when we know that a literal like `'1'` is definitely not going to have type `int?` at run time, and we don't have the same kind of guarantee with most user-written types. `final` classes can provide some exceptions, but in general we can have types like `C` above and `cast_always_throws` wouldn't apply.

> ```dart
> extension type ET(int number) implements int {}
> 
> void fn() {
>   '' as ET; // No lint, but throws at runtime
> }
> ```

That one could be flagged because of some kind of a `cast_always_throws` lint (if we get something like that). But `cast_into_extension_type` would not apply, because of the convention that "if you're an `int` then you're always an appropriate `ET`".

I think it's important to have the ability to allow transparent extension types to arise because of run-time type checks/tests, promotions, pattern matching, `on` clauses in a try/catch statement, and whatnot. We know that this isn't going to give us an inappropriate interpretation of the underlying representation object, and it's convenient to have this feature. For example:

```dart
extension type FancyString(String it) implements String {
  FancyString get fancify => FancyString('**$it**');
}

void main() {
  var xs = ['Hello', 'world!'];
  if (xs is List<FancyString>) { // Mass-promote all elements in `xs`.
    for (var x in xs) {
      print(x.fancify);
    }
  }
}
```

This is safer than `xs as List<FancyString>` exactly because the cast could be arbitrarily wrong (silently), but the promotion will only take us from a given type (like `List<String>`) to a subtype (like `List<FancyString>`), and not to an unrelated type.

**Comment by incendial** [comment]
> because most types can include values that are also subtypes of some other given type:

On paper that is true, but we have this rule available for ~2.5 year (maybe even more) and got only a few reports for that particular case. Of course, not all dart users use our tool, but in worst case people can just ignore the lint or rewrite the code, but get a correct warning for other (the majority) cases.

> '' as ET;
> "if you're an int then you're always an appropriate ET".

How is that appropriate if it throws at runtime? 🤔 I believe no user would want a runtime exception.

> I think it's important to have the ability to allow transparent extension types to arise because ...

Sure, but I still don't understand what practical difference does the transparency bring (sorry).
We can get the representation type either way and if it differs from the `implements T`, it's still related to it.

And I assume no user want to have code like that:

```dart
extension type FancyInt(int it) implements int {
  String get fancify => it.toString();
}

void main() {
  var xs = ['Hello', 'world!'];
  if (xs is List<FancyInt>) { // Never true
    for (var x in xs) {
      print(x.fancify);
    }
  }
}
```
as it simply never works (though in that particular case `fancify` is highlighted, but we can find another example).

**Comment by eernstg** [comment]
> How is that appropriate if it throws at runtime?

The property which is satisfied is that you can _trust_ the run-time type test/check: If it succeeds then you have an appropriate value of a transparent extension type `ET`, and if it fails then you don't (and the soundness property of the language ensures that you don't get the opportunity to access that inappropriate value with that typing at all).

With a non-transparent extension type like `Prime` you can perform the test, and it might succeed, and it might still be an inappropriate value of type `Prime` (like `4 as Prime`). In this case you cannot trust the run-time type test/check. It is still OK that it fails when it fails, but it may or may not be OK to use the value when the cast succeeds.

So you shouldn't cast or promote into the non-transparent type, you should call a constructor of `Prime`. The owners of the declaration of `Prime` are then able to write some code in the constructors of `Prime` that validates whatever they want (such that every expression of type `Prime` which has been obtained via a constructor invocation will be appropriate in the sense that the owners of this declaration have defined).

The practical difference that transparency offers is that it serves as a marker:

- This (transparent) extension type is appropriate for every object whose type is the representation type so we can rely on run-time type operations, vs. 
- This (non-transparent) extension type should be established by running a constructor.

> it simply never works (though in that particular case `fancify` is highlighted, but we can find another example).

Well, `fancify` is highlighted (and the invocation is a compile-time error) because there is no promotion, because `List<FancyInt>` is not a subtype of `List<String>`, because `FancyInt` isn't a subtype of `String`.

So that's working as intended: You don't get the type `FancyInt` for each element of `xs` because the run-time type test doesn't satisfy the requirements associated with promotions.

In both cases, we want to avoid the plain cast `e as ET` because it's dangerous. In the first case we promote a `List<String>` to a `List<FancyString>`, and it works fine because `FancyString` is transparent (and, in particular, a subtype of `String`), and in the second case we get a compile-time error because there is no promotion. All good. And the next thing we do is to fix the code where there is no promotion because missing promotions is rather often a hint that something is wrong. ;-)
